test/Oscoin/Test/Data/OscoinTx.hs

```diff
@@ -0,0 +1,183 @@
+module Oscoin.Test.Data.OscoinTx where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Address (Address)
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Crypto.PubKey (PublicKey)
+import           Oscoin.Data.Ledger
+import           Oscoin.Data.OscoinTx
+import           Oscoin.Test.Crypto
+import           Oscoin.Test.Util
+import           Test.Oscoin.P2P.Handshake (genKeyPair, genKeyPairPair)
+
+import qualified Crypto.Data.Auth.Tree as WorldState
+
+import           Test.Oscoin.Crypto.Address (genAddress)
+
+import           Test.Hedgehog.Extended
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+import           Hedgehog hiding (evalEither, forAll, (===))
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+
+tests :: forall c. Dict (IsCrypto c) -> [TestTree]
+tests Dict =
+    [ testGroup "Data.OscoinTx"
+        [ testProperty "Transfer balance"        (propTransferBalance @c Dict)
+        , testProperty "Apply transaction"       (propApplyTx         @c Dict)
+        ]
+    ]
+
+propApplyTx :: forall c. Dict (IsCrypto c) -> Property
+propApplyTx Dict = property $ do
+    (pk, pk') <- publicKeyPair @c
+
+    pk /== pk'
+
+    ws    <- forAll genWorldState
+    tx    <- mkTx <$> forAll (genRegisterProject pk)
+    alice <- forAll (genAddress pk')
+
+    -- Should fail because Alice doesn't have an account
+    applyTx tx alice ws ===  Left (ErrKeyNotFound (addressKey alice))
+
+    -- Give Alice an account
+    ws' <- evalEither $ creditAccount alice 1 ws
+
+    -- Should fail because the transaction nonce is incorrect
+    applyTx (tx { txNonce = 1 }) alice ws' ===  Left (ErrInvalidNonce 1)
```

Comment:
Would it make sense to split these tests up into one property each?

Reply:
I thought about it, but since they all share a common setup I thought it might be more messy ... :thinking: